rust/agama-server/src/questions/web.rs

```diff
@@ -0,0 +1,210 @@
+            }
+        }
+        Ok(result)
+    }
+
+    async fn create_generic_question(&self, path: &OwnedObjectPath) -> Result<Question, ServiceError> {
+        let dbus_question = GenericQuestionProxy::builder(&self.connection)
+            .path(path)?.cache_properties(zbus::CacheProperties::No).build().await?;
+        let result = Question {
+            generic: GenericQuestion {
+                id: dbus_question.id().await?,
+                class: dbus_question.class().await?,
+                text: dbus_question.text().await?,
+                options: dbus_question.options().await?,
+                default_option: dbus_question.default_option().await?,
+                data: dbus_question.data().await?
+            },
+            with_password: None
+        };
+
+        Ok(result)
+    }
+
+    async fn create_question_with_password(&self, path: &OwnedObjectPath) -> Result<Question, ServiceError> {
+        let dbus_question = QuestionWithPasswordProxy::builder(&self.connection)
+            .path(path)?.cache_properties(zbus::CacheProperties::No).build().await?;
+        let mut result = self.create_generic_question(path).await?;
+        result.with_password = Some(QuestionWithPassword{
+            password: dbus_question.password().await?
+        });
+
+        Ok(result)
+    }
+
+    pub async fn answer(&self, id: u32, answer: Answer) -> Result<(), ServiceError> {
+        let question_path = OwnedObjectPath::from(
+            ObjectPath::try_from(format!("/org/opensuse/Agama1/Questions/{}", id))
+                .context("Failed to create dbus path")?
+            );
+        if let Some(password) = answer.with_password {
+            let dbus_password = QuestionWithPasswordProxy::builder(&self.connection)
+            .path(&question_path)?.cache_properties(zbus::CacheProperties::No).build().await?;
+            dbus_password.set_password(password.password.as_str()).await?
+        }
+        let dbus_generic = GenericQuestionProxy::builder(&self.connection)
+            .path(&question_path)?.cache_properties(zbus::CacheProperties::No).build().await?;
+        dbus_generic.set_answer(answer.generic.answer.as_str()).await?;
+        Ok(())
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum QuestionsError {
+    #[error("Question service error: {0}")]
+    Error(#[from] ServiceError),
+}
+
+impl IntoResponse for QuestionsError {
+    fn into_response(self) -> Response {
+        let body = json!({
+            "error": self.to_string()
+        });
+        (StatusCode::BAD_REQUEST, Json(body)).into_response()
+    }
+}
+
+#[derive(Clone)]
+struct QuestionsState<'a> {
+    questions: QuestionsClient<'a>,
+}
+
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
+pub struct Question {
+    generic: GenericQuestion,
+    with_password: Option<QuestionWithPassword>,
+}
+
+/// Facade of agama_lib::questions::GenericQuestion
+/// For fields details see it.
+#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
```

Comment:
Why do you need a facade? If it is because of the `utoipa::ToSchema`, it is fine with me. I have derived the `utoipa::ToSchema` for some structs in lib and it feels wrong to me.

However, having to use a facade just because of our documentation library feels wrong too.

Reply:
No, it is not related to utoipa at all. It is because I do not want to break dbus code and I like more the approach with generic question and composition of question parts. Original code contain QuestionWithPassword that contain link to GenericQuestion, which I do not like much.

Reply:
Well, as soon as you do not break the D-Bus external API, you can refactor the internals if you wish. If you decide to keep the facade, please, write down the reason in the comment.

Reply:
well, I will write it to comment. In DBus we have attributes and in question is included also answer attribute. Which is not what I want to http API. I have there two parts: 1. question and 2. answer. So original one struct is split into two and the first one is used as output and the second as expected input. At least that is my idea. As said I will write it to comment